### ISSUE TITLE
CxSystem2 on the BSP

### ISSUE THREAD

**Issue by lbologna** [cnr-ibf-pa/hbp-bsp-issues#499]
## The CxSystem2 simulation framework is integrated in the BSP
 
Aspect | Detail
------------ | -------------
Summary | The CxSystem2 simulation framework is integrated in the BSP. The user can build networks, set network and simulation parameters, run simulations and download simulation results from the graphical user interface
Expert | @Vafa-Andalibi @henhok @sivanni 
Deadline | 31/01/2020
Dependencies | None
Test url | https://collab.humanbrainproject.eu/#/collab/66649/nav/501167
Notebooks | https://collab.humanbrainproject.eu/#/collab/67332/nav/457373
 
## List of additional/changed features 
The user can run example simulations by following the tutorial, change parameters of the example simulations, upload her/his configuration files, build a network from a template, run simulations, visualize neural activity, download simulation results.

### Tasks

- [x] The user runs an example simulation and downloads the results (a tutorial on how to use the application is linked in the app homepage)
- [x] The user uploads the configuration files and downloads the results (in order to get two example configuration files, one can select an example and download the configuration files via the "Export Parameters" and "Export Physiology" buttons).
- [x] Starting from a template, the user can modify/adapt the network and simulation parameters before running the simulation and downloading the results
- [x] Insert this Use Case item in the BSP Guidebook (a dedicated readthedocs page, which includes a tutorial on the app usage, is already available and can be linked on the BSP Guidebook)
- [x] Update documentation (Status box)
- [x] Update time information regarding STOP of simulation
- [x] Equations in the notebook

## Acceptance Criteria
- All the operations reported in the Task list are executed without errors
- All buttons work well
- The Use Case documentation is linked in the BSP's Guidebook
 
### Performance
At the moment the simulations can only run on the vm hosting the application and not on HBP related HPC systems

**Comment by alex4200** [comment]
@Vafa-Andalibi @henhok 

I tried the application and followed the tutorial, but when I click on 'Simulate' nothing happens. There is no reply from https://hbp-bsp-cxsys2.cineca.it/simulate, and after one minute I hit a timeout.

**Comment by Vafa-Andalibi** [comment]
@alex4200 thank you for your update. I added another notification for submission of the request. After you submit the request, you can check the status of the current simulation by clicking on the `simulation output` which shows you what's actually happening on the server regarding the simulations you have submitted. Also, if simulation is an array of simulations, you can check which part of the simulation is already done and the results are send by clicking on the `show workspace` button. Please let me know if you still experienced issues.

**Comment by alex4200** [comment]
@Vafa-Andalibi  Thanks for the reply, it is somewhat clearer now. However, I have some further questions/comments:

- It is hard to see for a user when the simulation has been finished. You need to check the "Simulation Output" or the "Show Workspace", but in the first view you see some not-intuitive log output, and in the second place you see just some files that are being created. To me it is really unclear if the simulation is running, if is has finished, if there was an error etc.
- In the documentation for Tutorial 3 you have a line saying:

`Here we use the curly brace syntax: set background_rate as {0|5|2}*Hz (corresponding to 0, 2, 4 Hz) and k as {1 & 3 & 5 & 7}. Now you have defined a 3x4 parameter search.`

    This is confusing. Do you use a vertical line of the '&' symbol?

- I have tried the `neurodynlib` tutorial and copied the first notebook `0 - Install` into one of my collabs. However, in the last cell (installation of CxSystem2) I get the following error

```
Obtaining file:///home/jovyan/CxSystem2
Exception:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/pip/req/req_install.py", line 1025, in check_if_exists
    self.satisfied_by = pkg_resources.get_distribution(str(no_marker))
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 558, in get_distribution
    dist = get_provider(dist)
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 432, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 968, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 859, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pip._vendor.pkg_resources.ContextualVersionConflict: (sympy 1.0 (/opt/conda/lib/python3.6/site-packages), Requirement.parse('sympy>=1.2'), {'brian2', 'Brian2'})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/opt/conda/lib/python3.6/site-packages/pip/commands/install.py", line 335, in run
    wb.build(autobuilding=True)
  File "/opt/conda/lib/python3.6/site-packages/pip/wheel.py", line 756, in build
    self.requirement_set.prepare_files(self.finder)
  File "/opt/conda/lib/python3.6/site-packages/pip/req/req_set.py", line 380, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "/opt/conda/lib/python3.6/site-packages/pip/req/req_set.py", line 521, in _prepare_file
    req_to_install.check_if_exists()
  File "/opt/conda/lib/python3.6/site-packages/pip/req/req_install.py", line 1036, in check_if_exists
    self.req.name
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 558, in get_distribution
    dist = get_provider(dist)
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 432, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 968, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 859, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pip._vendor.pkg_resources.ContextualVersionConflict: (sympy 1.0 (/opt/conda/lib/python3.6/site-packages), Requirement.parse('sympy>=1.2'), {'brian2', 'Brian2'})
```

   Seems some kind of version conflict. Do you have any ideas if that can be fixed?

**Comment by Vafa-Andalibi** [comment]
@alex4200 thanks for the great feedback. Regarding the first issue that you mentioned, I implemented a "status window" which gives you update about the submitted simulations. The window is by default on the bottom right and draggable and adjustable. 

Regarding your second comment: 
`{0|5|2}` will have the same behavior as `range(0,5,2)` in python where the first number is the start index, the second number is the end index and the last number is the step. Therefore it's equivalent to `0, 2, 4`. Another example would be `{1|2|0.1}` which is equivalent to `1.0, 1.1, 1.2 , ... , 1.9`. 
The other notation `&`, however, is literally equivalent to "and". So if you have `{0&5&2}` it's just `0` and `5` and `2` individually. 

When you use both of these at the same time their results would be combined together. Let's say for variable `A` we have `{0|5|2}` which is equivalent to `0, 2, 4` and for variable `B` we have `{0&5&2}` which is equivalent to `0, 5, 2`. In this case, our multidimensional array run would result in 9 simulations as follows: 

| sim idx  | A  |  B |
|---|---|---|
| 1  | 0  | 0  |
|  2 |  0 |  5 |
|  3 |  0 |  2 |
| 4  |  2 |  0 |
| 5  |  2 |  5 |
| 6  |  2 |  2 |
| 7  |  4 |  0 |
| 8  |  4 |  5 |
| 9  |  4 |  2 |

As another example, if for variable A we had `{0&2&4}` instead of  `{0|5|2}`, the result still would be identical to the table above. We will try to add more examples in the documentation to clarify this. 

Regarding your 3rd comment, I think the root of the problem might be an issue with some typo in requirements of `CxSystem2`. I updated that as well and it should be fixed. Please try again and let me know if it didn't work.

**Comment by alex4200** [comment]
@Vafa-Andalibi 

I strongly suggest to update the documentation to let users know that the format `{0,5,2}` means the same as `range(0,5,2)`. That is highly non-intuitive!

Furthermore, I still get almost the same error for the `0 - Install`step. I tried with a different user, but still get the same error!

```
Obtaining file:///home/jovyan/CxSystem2
Exception:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/pip/req/req_install.py", line 1025, in check_if_exists
    self.satisfied_by = pkg_resources.get_distribution(str(no_marker))
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 558, in get_distribution
    dist = get_provider(dist)
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 432, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 968, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 859, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pip._vendor.pkg_resources.ContextualVersionConflict: (Cython 0.25.2 (/opt/conda/lib/python3.6/site-packages), Requirement.parse('cython>=0.29'), {'brian2'})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/opt/conda/lib/python3.6/site-packages/pip/commands/install.py", line 335, in run
    wb.build(autobuilding=True)
  File "/opt/conda/lib/python3.6/site-packages/pip/wheel.py", line 756, in build
    self.requirement_set.prepare_files(self.finder)
  File "/opt/conda/lib/python3.6/site-packages/pip/req/req_set.py", line 380, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "/opt/conda/lib/python3.6/site-packages/pip/req/req_set.py", line 521, in _prepare_file
    req_to_install.check_if_exists()
  File "/opt/conda/lib/python3.6/site-packages/pip/req/req_install.py", line 1036, in check_if_exists
    self.req.name
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 558, in get_distribution
    dist = get_provider(dist)
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 432, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 968, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/opt/conda/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 859, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pip._vendor.pkg_resources.ContextualVersionConflict: (Cython 0.25.2 (/opt/conda/lib/python3.6/site-packages), Requirement.parse('cython>=0.29'), {'brian2'})
You are using pip version 9.0.3, however version 19.3.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

**Comment by Vafa-Andalibi** [comment]
Just to be clear `{0|5|2}` is considered equivalent as `range(0,5,2)` and `{0,5,2}` is not. Regardless, we will update the documentation to fully clarify this notation. 

I'll work on the installation issue with @henhok and @sivanni asap, and will write back as soon as the issue is fixed.

**Comment by Vafa-Andalibi** [comment]
@alex4200 the documentation of tutorial 3 is now updated and the parameter search is clarified now. 

Regarding the installation issue, could you please try again by removing the `--force-install` flag in the last block of the notebook? So it would be `!pip install -e .`

**Comment by sivanni** [comment]
Hi Alex. I started to replicate the issue you mention above "I have tried the neurodynlib tutorial and copied the first notebook 0 - Install into one of my collabs. However, in the last cell (installation of CxSystem2) I get the following error"

Which notebook do you mean? We have the following notebooks in our repo:
./cxsystem2/neurodynlib/tutorial/03_synaptic_currents.ipynb
./cxsystem2/neurodynlib/tutorial/02_export_import.ipynb
./cxsystem2/neurodynlib/tutorial/01_basic_use.ipynb
./cxsystem2/neurodynlib/tutorial/04_networks.ipynb

best regards,
Simo

Ok., Vafa let me know that it is the collaboratory notebook, ignore above.

**Comment by alex4200** [comment]
Hi, 

I think you missed a notebook:

![Screenshot 2020-01-17 at 07 33 10](https://user-images.githubusercontent.com/12729291/72589499-bbfba780-38fb-11ea-9c7f-413bfba64a6a.png)

**Comment by sivanni** [comment]
Thanks, Alex. Rather obvious. In the github repo we do not have this notebook zero, that is why I missed it, because I work locally. This notebook is for BSP integration only, and should cover everything you need to do for starting using it in BSP collab.  I wonder if the collab copy to own collabs should be included in this page. New user (such as me) may have difficulties finding that info.

**Comment by alex4200** [comment]
@sivanni  Hi, I am not sure I understand what you were explaining. I was just following the instructions given at https://cxsystem2.readthedocs.io/en/latest/tutorial/tutorial-4-neurodynlib.html where a link leads me to this collab page: https://collab.humanbrainproject.eu/#/collab/67332/nav/457373

Here I have the first notebook which is named '0 - Install' which I assume is a page to install everything I need. And since a new user (and myself) are not part of that Collab, I need to copy that notebook to one of my Collabs. Then in that Collab I ran the notebook to set everything up. 

But as mentioned before this is not working as I get some error. This error needs to be fixed before this new usecase can be added to the Brain Simulation Platform. 

I hope that clarifies it a bit - if not we can schedule a video call and talk in more detail abiut this use case and this ticket.

**Comment by henhok** [comment]
@alex4200 Thank you for all the testing and comments so far! 

I got similar (though not exactly the same) errors on my CSCS Jupyter environment. I tried the installation on a Jupyter notebook hosted at Jülich, which I had not used before, and it worked. I then asked the HBP support team to reset the container in CSCS, and the installation and the rest of it is working. 

So it seems that you can mess up your Jupyter environment so that it's no longer compatible with CxSystem. I didn't get these errors in December (before my paternity leave), and it's unclear to me what's causing them now.

In any case, resetting the Jupyter container seems to work. I don't know if there's a way to do it without contacting the support team? A virtual environment would be nice, but creating one, at least for a hosted notebook, seems to be non-trivial. 

I wonder how we should proceed. We probably can't ask everyone to reset their environments if they come across such dependency issues..?

**Comment by alex4200** [comment]
@henhok These jupyter container issues are well known 'shortcomings' of the Collaboratory - and as I am testing notebooks manually and automatically I have sometimes ask for a container reset multiple times - a day!

However, since there is the new Collab coming up and in user testing phase (see [HERE](https://jupyterhub.apps-dev.hbp.eu/)), which supposedly will handle issues like these much better, the (mid term) plan would be to migrate the notebooks to that new setup. 

But for the meantime (and the end of SGA2) we have to stick with what we have it seems so...

**Comment by henhok** [comment]
@alex4200 Oh dear! Good to know it's a common problem. I really hope the new Collab resolves these issues. 

So is it ok if we add a text asking users to request a container reset if they get dependency issues? I mean for the current Collab.

I can't access the new Collab ("Application is not available"), but I'm happy to migrate notebooks there when it's available.

**Comment by alex4200** [comment]
@henhok Yes, maybe you can add a (red) box with some comment to that use case regarding a container reset.

**Comment by lbologna** [comment]
Dear All,
besides the work that you are doing on the Jupyter Notebook for the CxSystem2 (which can later be added as a new use case in the Platform), I believe the web app is ready to be integrated into the BSP so that users can start cloning it to their Collabs.

There are a couple of (minor) things that should be corrected/changed in my opinion:
1) the "Status of Simulations" panel should have a x (cross) on the top-right corner so that it can be closed by the user if he wants to, and it should be reopened when the "Simulate" button is pressed.
2) when stretching the browser window, the layout of some buttons on the top menu (and of the above mentioned status panel) is disrupted (see attached figures Layout_01.png and Layout_02.png). Is it possible to correct this?

Given that Alex has already tested the web app, the integration will be probably possible in a few days (hopefully before the summit). I can modify the use case config file so that the use case appears in the list. For this I will need an icon. If it is fine with you, I can use the image that you used in the README file of the Github repository (that I attach herewith; file: icon.png). If you want to use some other image, please send it to me.
Thank you.

![Layout_01](https://user-images.githubusercontent.com/9949990/73259014-2ac3e500-41c7-11ea-9f72-adab556a6846.png)





![Layout_02](https://user-images.githubusercontent.com/9949990/73259017-2ac3e500-41c7-11ea-9e22-0ffb9a27f4c5.png)





![icon](https://user-images.githubusercontent.com/9949990/73259013-2a2b4e80-41c7-11ea-9f55-46c11aac0354.png)

**Comment by alex4200** [comment]
The webapplication for the CxSystem2 **does not work** for me. I follow the tutorial and start the example simulation, but it never seems to end (already running for more than 90 minutes). 
_This has to be fixed and retested properly before the application can go into production._

**Comment by lbologna** [comment]
@alex4200 I see. I understood from recent chats that the issues to be solved only concerned the notebook. I will  give a try to the web-app as well and give you a feedback. Thanks.

**Comment by alex4200** [comment]
I forgot to mention: The notebooks seem to work. I was able to run all of them today.

I have to add: I think the shown equation should be rendered better. I see the following image:

![Selection_971](https://user-images.githubusercontent.com/12729291/73277761-e9453100-41ea-11ea-8185-7af220658a41.png)

This does not look correct.

**Comment by Vafa-Andalibi** [comment]
@lbologna I added the close button and I'm working on fixing the alignment. There was a bug in array_run in the VM caused by the virtualenv which is now fixed and Alex has already tested the CobaEIF example.

**Comment by Vafa-Andalibi** [comment]
@lbologna I temporarily fixed that button alignment issue by displaying a warning about browser being too small. as soon as they increase their browser size a little bit larger everything will automatically show up.
![resolution](https://user-images.githubusercontent.com/9551535/73323365-3ccc7480-4215-11ea-922d-c06ff86a0be8.png)

**Comment by Vafa-Andalibi** [comment]
@alex4200 the documentation ([here](https://cxsystem2.readthedocs.io/en/latest/tutorial/tutorial-1-run-example.html#)) now contains information about the sim status window. 
Also the simulation status window now contains the correct time information for start and end: 
![image](https://user-images.githubusercontent.com/9551535/73324801-3ab8e480-421a-11ea-8406-02a864a942bc.png)

**Comment by alex4200** [comment]
@Vafa-Andalibi Thanks for the changes, looks good to me. A re-run of the simulation also did work.

**Comment by lbologna** [comment]
Hello,
@Vafa-Andalibi I checked the layout/window resizing and worked well for me.

@alex4200 was the simulation that worked for you concerning the web app? can I proceed with the BSP integration?

@Vafa-Andalibi @henhok @sivanni @mmigliore  what do you think the Use Case title should be?
Do you think something like: "Configure and run a small cortical circuit with the CxSystem2 simulation framework?" could work?

Thank you all for your feedback.

**Comment by sivanni** [comment]
Thanks Luca for the suggestion. I suggest tiny modification:
"Configure and run arrays of simplified cortical circuits with the CxSystem2 simulation framework."
What do @Vafa-Andalibi and @henhok think? Ok, for @mmigliore ?

**Comment by henhok** [comment]
@sivanni @lbologna @Vafa-Andalibi Configure and run a simplified cortical circuit with the CxSystem2 simulation framework? “Small” and “array” are confusing to me :)

**Comment by sivanni** [comment]
One of the critical benefits of CxSystem is to enable parallel searches of network structures. I currently run 200-1000 network models in parallel -- part of it for experimental conditions and replications, and part for parameter search. This is one of the unique strengths of CxSystem. In addition, Brian supports customizing all parts of the model, and this has been pivotal in design. Thus, I suggest we somehow bring up the support for parallel searches in the Use Case. How about  "Construct simplified cortical circuits and run multiple models in parallel with the CxSystem2 simulation framework."

**Comment by sivanni** [comment]
Luca asked for shorter title, Henri and I thorught of  "Construct and simulate  simplified cortical circuits" @Vafa-Andalibi , OK ?

**Comment by Vafa-Andalibi** [comment]
@sivanni Definitely

**Comment by alex4200** [comment]
@Simo Vanni @Henri Hokkanen @Vafa Can you recommend someone who is familiar with the scientific content and context of that usecase, but did not actually work on it? I would to have someone with background knowledge to perform a user testing on that use case in order to see if it does really work and is useable.

Also an entry in the Guidebook is missing with some kind of documentation (can be very short, referencing the tutorial that comes with the use case). (See https://github.com/lbologna/hbp-sp6-guidebook for the repository).

**Comment by sivanni** [comment]
Dear Alex,

As far as I know, we have no outside users yet. Only people outside our group familiar with our use case are Carmen and Luca from Palermo. Alternatively, I will have a summerstudent ariving in May and she will be using the simulator without contributing to its code. Please let me know whom you would prefer.

We will work for the entry for the guidebook, and get back to you.

**Comment by alex4200** [comment]
@lbologna volunteers to take a look at the use case.

**Comment by sivanni** [comment]
Thanks @lbologna ! Can you @alex4200 please give some specs for the requested user testing for Luca to proceed.

**Comment by lbologna** [comment]
Hi all,

@alex4200 I am going on with the testing: I am mostly testing parameter setting and visualizing/monitoring results. This should be finalized early this week.
@sivanni I am in contact with Alex on Skype and slack for any specific issue during testing.

I'll keep you posted.
Thank you

**Comment by lbologna** [comment]
Hello,

@alex4200 I finalized a test with complete network setting and running. It worked for me.
If it's fine with you I will insert it in the usecases interface.
@sivanni , I can take care of adding the link to the documentation/tutorial (already existing in readthedocs) in the BSP Guidebook

**Comment by alex4200** [comment]
@lbologna  Yes sure, please go ahead.

**Comment by lbologna** [comment]
Hello @alex4200,
I just generated a pull request with the newly added use case.
Could you please have a look?
Thank you

**Comment by alex4200** [comment]
@lbologna I tested the implementation on STAGE and everything seems to work so far. 

Two items to be done:
- [x] Documentation on the guidebook
- [ ] What does the open point mean "equations in the notebook"?

**Comment by lbologna** [comment]
> @lbologna I tested the implementation on STAGE and everything seems to work so far.
> 
> Two items to be done:
> 
>     * [x]  Documentation on the guidebook
> 
>     * [ ]  What does the open point mean "equations in the notebook"?

Hello @alex4200,
thanks for checking. I finalized the Guidebook page:
https://lbologna.github.io/hbp-sp6-guidebook/index.html
If it's fine with you I will move it to prod as soon as the use case is available in the BSP.

As for the "equations in notebook" point, I think it refers to the test collab that was originally created with example notebooks:
https://collab.humanbrainproject.eu/#/collab/67332/nav/457373

I see some equations there so maybe this point is finalized (I checked it). It doesn't affect the integration in the BSP anyway so I believe it is not a blocking point.
Thanks.

**Comment by sivanni** [comment]
I checked Luca's txt for the Guidebook and made some additions. I could not, however, push them and did not find a straightforward way of creating a pull request. I will e-mail the changed .rst text to Luca for incorporation into the Guidebook. Thanks in advance!

**Comment by antonelepfl** [comment]
Luca will add the documentation on the Guidebook but this UC is in PROD

**Comment by lbologna** [comment]
Just added the documentation in the Guidebook (Prod)
